<h1>Help</h1>

<h2>Genome browser</h2>
<p>
	FoldAtlas includes a genome browser called d3nome, which shows intron and exon boundaries for each transcript, along with features that are near to the transcript of interest. This facilitates genome exploration and identification of interesting transcripts.
</p>

[show a picture of the genome browser here]

[show bullet list describing each element]

<h2>Searching for transcripts</h2>
<h3>Go to a known transcript ID</h3>
<p>
	To search for a known transcript identifier (e.g. the TAIR identifier "AT1G01010.1"), hit "Search", followed by the "Transcript ID" tab. Next, the transcript transcript ID should be entered and the "Search" button pressed. This will search the FoldAtlas database for the transcript ID, and jump to the profile page of the first match.
</p>
<img class="help-example-image" src="{{ settings.static_base_url }}/assets/help/search-transcript-id.png" />

<h3>Find high coverage transcripts</h3>
<p>
	An issue with genome-scale RNA structure probing experiments is that many transcripts have only a small number of reads mapped to them, i.e. are low-coverage. We have produced an interface that lists transcripts in order of coverage. This can be reached by pressing Search and then choosing the Coverage tab. Transcripts with the best coverage are listed first, so that high-coverage transcripts can be found easily. The right column indicates transcripts where <i>in vivo</i> structures are available.
</p>
<img class="help-example-image" src="{{ settings.static_base_url }}/assets/help/search-coverage.png" />

<h2>Structure views</h2>
<p>
	FoldAtlas holds RNA secondary structure predictions generated using the RNAstructure package [REF]. Predictions are divided into two categories: <i>in vivo</i>, and <i>in silico</i>. Each structure prediction has an estimated free energy term associated, in kcal/mol.

<h3><i>In silico</i> predictions</h3>
<p>
	All transcript sequences have <i>in silico</i> structures, which have been generated from the RNA sequence alone, without constraints.
</p> 

<h3><i>In vivo</i> predictions</h3>
<p>
	The <i>in vivo</i> structures have been generated by constraining the prediction method with DMS reactivity values, generated in a previous genome-scale screen [REF]. Constraining predictions using experimental structure probing data has been shown to improve the predictive accuracy [REF], hence <i>in vivo</i> predictions are inferred to be of better accuracy than their <i>in silico</i> counterparts.
</p><p>
	It is important to note that not all transcripts have <i>in vivo</i> predictions, since many transcripts have low read sequence coverage, corresponding to low RNA abundance. The proportion of transcripts with <i>in vivo</i> predictions will improve in future FoldAtlas iterations, as read depth from experiments increases.
<p>

<h3>PCA plots for structural similarity visualisation</h3>
<img class="help-example-image-small" src="{{ settings.static_base_url }}/assets/help/pca.png" />
<p>
	When the RNAstructure prediction method is run on a transcript sequence, up to 20 low free energy structures are generated. In FoldAtlas, the similarity between these structures is visualised using Principle Component Analysis (PCA). Each dot on the plot is a structure, and similar structures are located in close proximity to one another.
</p><p>
	Clicking a dot shows detailed information about the structure. A red dot colour indicates lower free energy, i.e. a more favourable structure. The minimum free energy (MFE) structure is visualised by default.
</p><p>
	The method for using PCA with RNA structures has been described in an earlier work [REF], and involves generating a vector of length <i>n</i>, where <i>n</i> is the number of nucleotides. The value at each position in the vector is set to either 0 if the corresponding base is unpaired, or 1 if paired. The vectors are then entered into the PCA method [REF], with each vector position as a dimension.
</p>

<h3>Circle plots for base pairing visualisation</h3>
<img class="help-example-image-small" src="{{ settings.static_base_url }}/assets/help/circle.png" />
<p>
	The circle plot can be used to intuitively visualise base pairing in a predicted RNA structure [REF]. Bases are arranged in a circle according to their sequence position, and lines drawn between bases indicate pairing. The free energy estimate is displayed in the top right corner.
</p>

<h3>Structure diagrams for fold visualisation</h3>
<img class="help-example-image-small" src="{{ settings.static_base_url }}/assets/help/structure-diagram.png" />
<p>
	FoldAtlas also visualises structures using a fold diagram. The diagram is generated using the <i>ViennaRNA</i>[REF] package and then displayed in the browser using the <i>forna</i>[REF] JS module. Clicking "Enable interaction" allows the structure diagram to be rearranged by clicking and dragging bases. A free energy estimate is displayed in the top right corner.
</p><p>
	The nucleotides are coloured according to their reactivity. Green nucleotides represent weak or no reactivity, orange denotes moderate reactivity, whilst red indicates strong reactivity. White nucleotides indicate that reactivity is irrelevant. G and U nucleotides are always coloured white because we are displaying DMS reactivity data, and DMS only reacts with unpaired C and A nucleotides.
</p>

<h3>Structure data download</h3>
<img class="help-example-image" src="{{ settings.static_base_url }}/assets/help/structure-download.png" />
<p>
	The "Download" button illustrated above generates a tab-delimited text file containing up to 20 <i>in silico</i> structures, along with up to 20 <i>in vivo</i> structures if coverage is sufficient. This snippet illustrates the first 4 bases of an <i>in silico</i> structure:

<pre class="help-pre">
285	In silico	Col_0	AT1G01010.1	-471.9	-5.32598	-3.91789	A	1	0
285	In silico	Col_0	AT1G01010.1	-471.9	-5.32598	-3.91789	A	2	0
285	In silico	Col_0	AT1G01010.1	-471.9	-5.32598	-3.91789	A	3	1626
285	In silico	Col_0	AT1G01010.1	-471.9	-5.32598	-3.91789	T	4	1625
</pre>

<p>
	The columns represent, from left to right:

	<ol>
		<li>A unique identifier for the structure, used internally by FoldAtlas</li>
		<li>Description of the prediction method; can be <i>in silico</i> or <i>in vitro</i>.</li>
		<li>The strain from which the RNA sequence originated. Currently this is Col_0, the <i>A. thaliana</i> reference strain.</li>
		<li>The TAIR transcript indentifier.</li>
		<li>Free energy estimate, in kcal/mol, generated by <i>RNAstructure</i>.</li>
		<li>Principle Component 1, calculated during PCA. [INTERNAL REF]</li>
		<li>Principle Component 2, calculated during PCA. [INTERNAL REF]</li>
		<li>Nucleotide letter.</li>
		<li>Position of this nucleotide.</li>
		<li>The position that this nucleotide pairs to. 0 indicates no base pairing.</li>
	</ol>
</p>


[text file snippet]

[description]




<h2>References</h2>



<!--

Genome browser 
	Click and drag to move 
	Use menu to select chromosome
	Click labels to see details

DMS reactivities
	Source of data
	Spliced transcripts
	Description of the normalisation method
	Detailed view

(Ribosome profiling - later)

Downloading data
	Explain Download All SQL format, including ER diagram
	Explain structure download format - dotbracket is bad since it can't handle pseudoknots
	Explain DMS reactivity format

-->