<h1>Help</h1>

<h2>Contents</h2>

<ul>
    <li><a href="#contents_genome-browser">Genome browser</a></li>
    <li>
        <a href="#contents_search">Searching for transcripts</a>
        <ul>
            <li><a href="#contents_search-transcript-id">Go to a known transcript ID</a></li>
            <li><a href="#contents_search-coverage">Find high coverage transcripts</a></li>
        </ul>
    </li>
    <li>
        <a href="#contents_structure">Structure views</a>
        <ul>
            <li><a href="#contents_structure-in-silico"><i>In silico</i> predictions</a></li>
            <li><a href="#contents_structure-in-vivo"><i>In vivo</i> predictions</a></li>
            <li><a href="#contents_structure-pca">PCA plots for structural similarity visualisation</a></li>
            <li><a href="#contents_structure-circle">Circle plots for base pairing visualisation</a></li>
            <li><a href="#contents_structure-diagrams">Structure diagrams for fold visualisation</a></li>
            <li><a href="#contents_structure-download">Structure data download</a></li>
        </ul>
    </li>
    <li>
        <a href="#contents_reactivity">Reactivity measurements</a>
        <ul>
            <li><a href="#contents_reactivity-chart">Reactivity charts</a></li>
            <li><a href="#contents_reactivity-download">Downloading probing data</a></li>
        </ul>
    </li>
    <li>
        <a href="#contents_database">Database download</a>
        <ul>
            <li><a href="#contents_database-structural">Structural entities</a></li>
            <li><a href="#contents_database-genomic">Genomic entities</a></li>
        </ul>
    </li>
    <li><a href="#contents_references">References</li></a></li>
</ul>

<h2 id="contents_genome-browser">Genome browser</h2>
<p>
    FoldAtlas includes a simple genome browser called <i>d3nome</i>, which shows intron and exon boundaries for each transcript, along with features that are near to the transcript of interest. This facilitates genome exploration and identification of interesting transcripts.
</p>

<div class="help-example-image">
    <img style="width: 700px;" src="{{ settings.static_base_url }}/assets/help/d3nome.png" />

    <div class="help-marker" style="top: 20px; left: 35px;">1</div>
    <div class="help-marker" style="top: 178px; left: 108px;">2</div>
    <div class="help-marker" style="top: 77px; left: 310px;">3</div>
    <div class="help-marker" style="top: 219px; left: 339px;">4</div>
    <div class="help-marker" style="top: 219px; left: 566px;">5</div>
</div>

<p>
    <ol>
        <li>
            <b>Navigation bar:</b>
            This shows the chromosome coordinates, and can be clicked and dragged to change the region of the chromosome displayed.
        </li><li>
            <b>Viewing area:</b>
            Clicking and dragging this region will change the region of the chromosome displayed. The mouse wheel can also be used here to zoom in and out.
        </li><li>
            <b>Gene:</b>
            Genes are marked using boxes, with curved lines indicating introns. Filled boxes indicate coding sequence (CDS), whilst unfilled boxes are untranslated regions (UTRs). Green genes are transcribed on the forward strand, whilst red genes are transcribed on the reverse strand. Clicking the gene label will show detailed information about the gene.
        </li><li>
            <b>Resize bar:</b>
            Can be dragged downward to increase the view region.
        </li><li>
            <b>Chromosome selector:</b>
            This menu can be used to change the chromosome displayed.
        </li>
    </ol>
</p>

<h2 id="contents_search">Searching for transcripts</h2>
<h3 id="contents_search-transcript-id">Go to a known transcript ID</h3>
<p>
    To search for a known transcript identifier (e.g. the TAIR 

        (<a class="ref-link" href="#lamesch_2011">Lamesch <i>et al</i>., 2011</a>) 

    identifier "AT1G01010.1"), <b>Search</b> in the title bar should be pressed, followed by the <b>Transcript ID</b> tab. Next, the transcript transcript ID should be entered and the <b>Search</b> button pressed. This will search the FoldAtlas database and jump to the profile page of the first match.
</p>
<img class="help-example-image" src="{{ settings.static_base_url }}/assets/help/search-transcript-id.png" />

<h3 id="contents_search-coverage">Find high coverage transcripts</h3>
<p>
    An issue with genome-scale RNA structure probing experiments is that many transcripts have only a small number of reads mapped to them, i.e. are low-coverage. We have produced an interface that lists transcripts in order of coverage. This can be reached by pressing <b>Search</b> and then choosing the <b>Coverage</b> tab. Transcripts with the best coverage are listed first, so that high-coverage transcripts can be found easily. The right column indicates transcripts where <i>in vivo</i> structures are available.
</p>
<img class="help-example-image" src="{{ settings.static_base_url }}/assets/help/search-coverage.png" />

<h2 id="contents_structure">Structure views</h2>
<p>
    FoldAtlas holds RNA secondary structure predictions generated using the RNAstructure package 

        (<a class="ref-link" href="#mathews_2004">Mathews <i>et al</i>., 2004</a>)

    . Predictions are divided into two categories: <i>in vivo</i>, and <i>in silico</i>. Each structure prediction has an estimated free energy term associated, in kcal/mol.

<h3 id="contents_structure-in-silico"><i>In silico</i> predictions</h3>
<p>
    All transcript sequences have <i>in silico</i> structures, which have been generated from the RNA sequence alone, without constraints.
</p> 

<h3 id="contents_structure-in-vivo"><i>In vivo</i> predictions</h3>
<p>
    The <i>in vivo</i> structures have been generated by constraining the prediction method with dimethyl sulfate (DMS) reactivity values, generated in a previous genome-scale screen 

        (<a class="ref-link" href="#ding_2014">Ding <i>et al</i>., 2014</a>) 

    . Constraining predictions using experimental structure probing data has been shown to improve the predictive accuracy 

        (<a class="ref-link" href="#deigan_2008">Deigan <i>et al</i>., 2008</a>) 

    , hence <i>in vivo</i> predictions are inferred to be of better accuracy than their <i>in silico</i> counterparts.
</p><p>
    It is important to note that not all transcripts have <i>in vivo</i> predictions, since many transcripts have low read sequence coverage, corresponding to low RNA abundance. The proportion of transcripts with <i>in vivo</i> predictions will improve in future FoldAtlas iterations, as read depth from experiments increases.
<p>

<h3 id="contents_structure-pca">PCA plots for structural similarity visualisation</h3>
<img class="help-example-image-small" src="{{ settings.static_base_url }}/assets/help/pca.png" />
<p>
    When the RNAstructure prediction method is run on a transcript sequence, up to 20 low free energy structures are generated. In FoldAtlas, the similarity between these structures is visualised using Principle Component Analysis (PCA). Each dot on the plot is a structure, and similar structures are located in close proximity to one another.
</p><p>
    Clicking a dot shows detailed information about the structure. A red dot colour indicates lower free energy, i.e. a more favourable structure. The minimum free energy (MFE) structure is visualised by default.
</p><p>
    The method for using PCA with RNA structures has been described in an earlier work 

        (<a class="ref-link" href="#halvorsen_2010">Halvorsen <i>et al</i>., 2010</a>)

    , and involves generating a vector of length <i>n</i>, where <i>n</i> is the number of nucleotides. The value at each position in the vector is set to either 0 if the corresponding base is unpaired, or 1 if paired. The vectors are then entered into the PCA method, with each vector position as a dimension.
</p>

<h3 id="contents_structure-circle">Circle plots for base pairing visualisation</h3>
<img class="help-example-image-small" src="{{ settings.static_base_url }}/assets/help/circle.png" />
<p>
    The circle plot can be used to intuitively visualise base pairing in a predicted RNA structure. Bases are arranged in a circle according to their sequence position, and lines drawn between bases indicate pairing. The free energy estimate is displayed in the top right corner.
</p>

<h3 id="contents_structure-diagrams">Structure diagrams for fold visualisation</h3>
<img class="help-example-image-small" src="{{ settings.static_base_url }}/assets/help/structure-diagram.png" />
<p>
    FoldAtlas also visualises structures using a fold diagram. The diagram is generated using the <i>ViennaRNA</i>

        (<a class="ref-link" href="#hofacker_2003">Hofacker <i>et al</i>., 2003</a>)

    package and then displayed in the browser using the <i>forna</i> JS module. Clicking </b>Enable interaction</b> allows the structure diagram to be rearranged by clicking and dragging bases. A free energy estimate is displayed in the top right corner.
</p><p>
    The nucleotides are coloured according to their reactivity. Green nucleotides represent weak or no reactivity, orange denotes moderate reactivity, whilst red indicates strong reactivity. White nucleotides indicate that reactivity is irrelevant. G and U nucleotides are always coloured white because we are displaying DMS reactivity data, and DMS only reacts with unpaired C and A nucleotides.
</p>

<h3 id="contents_structure-download">Structure data download</h3>
<img class="help-example-image" src="{{ settings.static_base_url }}/assets/help/structure-download.png" />
<p>
    The <b>Download</b> button illustrated above generates a tab-delimited text file containing up to 20 <i>in silico</i> structures, along with up to 20 <i>in vivo</i> structures if coverage is sufficient. The following snippet illustrates 10 bases from an <i>in silico</i> structure.
</p>

<pre class="help-pre">
<u>1       2          3           4      5        6          7   8       9   </u>
...
5273    In vivo   AT1G01010.1 -850.4  9.10729 -1.76294    U   1294    130
5273    In vivo   AT1G01010.1 -850.4  9.10729 -1.76294    C   1295    0
5273    In vivo   AT1G01010.1 -850.4  9.10729 -1.76294    C   1296    126
5273    In vivo   AT1G01010.1 -850.4  9.10729 -1.76294    A   1297    125
5273    In vivo   AT1G01010.1 -850.4  9.10729 -1.76294    C   1298    124
5273    In vivo   AT1G01010.1 -850.4  9.10729 -1.76294    C   1299    0
5273    In vivo   AT1G01010.1 -850.4  9.10729 -1.76294    A   1300    121
5273    In vivo   AT1G01010.1 -850.4  9.10729 -1.76294    A   1301    120
5273    In vivo   AT1G01010.1 -850.4  9.10729 -1.76294    C   1302    119
5273    In vivo   AT1G01010.1 -850.4  9.10729 -1.76294    A   1303    0
...
</pre>

<p>
    The columns represent, from left to right:

    <ol>
        <li>A unique identifier for the structure, used internally by FoldAtlas</li>
        <li>Description of the prediction method; can be <i>in silico</i> or <i>in vivo</i>. The latter indicates that predictions were made using constraints from a DMS chemical probing experiment.</li>
        <li>The TAIR transcript indentifier.</li>
        <li>Free energy estimate, in kcal/mol, generated by <i>RNAstructure</i>.</li>
        <li>Principle Component 1, calculated during PCA.</li>
        <li>Principle Component 2, calculated during PCA.</li>
        <li>Nucleotide letter.</li>
        <li>Position of this nucleotide.</li>
        <li>The position that this nucleotide pairs to. 0 indicates no base pairing.</li>
    </ol>
</p>

<h2 id="contents_reactivity">Reactivity measurements</h3>
<p>
    FoldAtlas includes a module that plots DMS reactivity charts and allows them to be downloaded. At present, all FoldAtlas structure probing measurements are from DMS experiments. In the future, we will also add selective 2'-hydroxyl acylation analyzed by primer extension (SHAPE) measurements. 
</p><p>
    The DMS reactivities stored in FoldAtlas are normalised as described in 
        <a class="ref-link" href="#ding_2014">Ding et al. (2014)</a>
    , under "Methods", "Determination and normalization of DMS reactivity".
</p>

<h3 id="contents_reactivity-chart">Reactivity charts</h3>
<p>
    The chart area includes an overview mode, which displays reactivities for the entire transcript on a single line. This can be seen at the top of the image below. Underneath the overview is a <b>More detail</b> button, which shows the transcript on multiple lines, with 80 nucleotides per line. In this detailed view, G and U nucleotide labels are coloured grey to indicate that they do not react with DMS.
</p>
<img class="help-example-image" src="{{ settings.static_base_url }}/assets/help/measurements-chart.png" />

<h3 id="contents_reactivity-download">Downloading probing data</h3>
<img class="help-example-image" src="{{ settings.static_base_url }}/assets/help/probing-download.png" />
<p>
   <b>Download Normalised</b> can be used to obtain a tab-delimited text file containing the normalised DMS reactivity measurements for a given transcript. The following snippet illustrates 11 bases from the AT1G01620.1 transcript:
</p>

<pre class="help-pre">
<u>1   2   3       </u>
75  U   NA
76  G   NA
77  U   NA
78  A   0.492188
79  G   NA
80  A   0.767697
81  A   0.940973
82  A   0.45815
83  A   0.0
84  G   NA
85  A   0.823105
</pre>

<p>
    The columns include, from left to right:
    <ol>
        <li>The nucleotide position. Numbers start from 1.</li>
        <li>The base type letter.</li>
        <li>The reactivity. 0 indicates no reactivity whilst 1 or more indicates the strongest reactivity. NA indicates a G or U nucleotide, where DMS reactivities are not relevant.</li>
    </ol>
</p>

<p>
   <b>Download Raw</b> can be used to obtain a text file containing raw adduct counts for DMS and control lanes. This snippet illustrates 11 bases from the AT1G01620.1 transcript:
</p>

<pre class="help-pre">
<u>1   2   3    4    </u>
75  U   145  176
76  G   80   0
77  U   292  2
78  A   125  447
79  G   148  157
80  A   92   704
81  A   78   1195
82  A   50   182
83  A   66   0
84  G   192  499
85  A   16   262
</pre>

<p>
    The columns include, from left to right:
    <ol>
        <li>The nucleotide position. Numbers start from 1.</li>
        <li>The base type letter.</li>
        <li>Adduct count from the control lane.</li>
        <li>Adduct count from the DMS lane.</li>
    </ol>
</p>

<h2 id="contents_database">Database download</h2>

The entire FoldAtlas MySQL database can be downloaded as an SQL file by using the <b>Download All</b> button:

<img class="help-example-image" src="{{ settings.static_base_url }}/assets/help/all-download.png" />

The following Entity-Relationship (ER) diagram describes the FoldAtlas database:

<img class="help-example-image-full" src="{{ settings.static_base_url }}/assets/help/er-diagram.png" />

<h3 id="contents_database-structural">Structural entities</h3>

<p>
    Structural entities include RNA structure prediction data, along with the probing data that were used to generate them.
    <ul>
        <li>
            <b>Structure Prediction Run:</b>
            This represents an RNA structure prediction run. At present there is just one of these, generated using the RNAstructure 

                (<a class="ref-link" href="#mathews_2004">Mathews <i>et al</i>., 2004</a>)
                
            method.
        </li><li>
            <b>Structure:</b>
            Represents a single structure that has been predicted as part of a structure prediction run. This is generated from a single transcript sequence. Includes a free energy estimate, data for PCA plots, and a string containing tab delimited pairing data.
        </li><li>
            <b>Nucleotide Measurement Run:</b>
            Represents a genome-scale experiment that involves taking measurements at specific nucleotide positions. At present this is used to represent DMS reactivity data. In the future, data from SHAPE and ribosome profiling experiments will also be represented by this entity.
        </li><li>
            <b>Raw Reactivities:</b>
            This represents raw adduct counts, generated from mapping reads in an RNA structure-seq experiment. Includes a text field containing the tab delimited adduct counts.
        </li><li>
            <b>Nucleotide Measurement Set:</b>
            Represents normalised reactivities, generated from raw reactivities. Includes a property to store a per-nucleotide coverage value for the entire transcript, and a string containing the individual measurements.
        </li>
    </ul>
</p>

<h3 id="contents_database-genomic">Genomic entities</h3>

<p>
    Other FoldAtlas tables deal with representing genomic data structures. These are used by the genome browser and also to generate RNA sequences. These cover gene and transcript identifiers. Genomic entities include:
    <ul>
        <li>
            <b>Strain:</b>
            A strain represents a single sequenced accession of a particular organism. At present, the only strain stored in FoldAtlas is Col0, which represents the <i>A. thaliana</i> reference genome. Future versions of FoldAtlas will include additional strains and organisms.
        </li><li>
            <b>Chromosome:</b>
            All FoldAtlas sequences originate from chromosome entities. Each chromosome is associated with a strain and a chromosome ID, which is a generic identifier shared across strains. The sequence property stores the entire chromosome sequence.
        </li><li>
            <b>Feature:</b>
            A feature describes start and stop boundaries within a chromosome that correspond to coding sequences (CDSs), untranslated regions (UTRs) or introns. By combining  features and chromosome sequences, it is possible to find the sequence of a transcript.
        </li><li>
            <b>Gene Location:</b>
            A gene location contains redundant information about the start and stop boundaries of a gene with respect to the genome sequence of a particular strain. These data are used to quickly calculate gene lengths, and to visualise gene boundaries in the integrated <i>d3nome</i> genome browser.
        </li><li>
            <b>Gene:</b>
            FoldAtlas gene entities are generic, i.e. are shared between different strains. A gene can therefore have multiple strain-specific locations, and can also be associated with multiple transcripts, representing alternative splice forms.
        </li><li>
            <b>Transcript:</b>
            The transcript is a generic entity that can be shared amongst multiple strains, with the transcript sequence represented by chromosome sequence and strain-specific features that cover the chromosome sequence.
        </li>
    </ul>
</p>

<h2 id="contents_references">References</h2>

<ul>
    <li id="deigan_2008" class="ref">
        Deigan, K. <i>et al</i>., 2008. Accurate SHAPE directed RNA structure determination. PNAS 106, 97-102.
    </li>
    <li id="ding_2014" class="ref">
        Ding, Y. <i>et al</i>., 2014. In vivo genome-wide profiling of RNA secondary structure reveals novel regulatory features. Nature 505, 696–700.
    </li>
    <li id="halvorsen_2010" class="ref">
        Halvorsen, M. <i>et al</i>., 2010. Disease-associated mutations that alter the RNA structural ensemble. PLoS Genetics 6, e1001074.
    </li>
    <li id="hofacker_2003" class="ref">
        Hofacker, L., 2003. Vienna RNA secondary structure server. Nucl. Acids Res. 31, 3429–3431.
    </li>
    <li id="lamesch_2011" class="ref">
        Lamesch, P. <i>et al</i>., 2011. The Arabidopsis Information Resource (TAIR): improved gene annotation and new tools. Nucleic Acids Research 40, D1202–1210.
    </li>
    <li id="mathews_2004" class="ref">
        Mathews, D. <i>et al</i>., 2004. Incorporating chemical modification constraints into a dynamic programming algorithm for prediction of RNA secondary structure. PNAS 101, 7287–7292.
    </li>
</ul>

<!--

Genome browser 
    Click and drag to move 
    Use menu to select chromosome
    Click labels to see details

References

DMS reactivities
    Source of data
    Spliced transcripts
    Description of the normalisation method
    Detailed view

(Ribosome profiling - later)

Downloading data
    Explain Download All SQL format, including ER diagram
    Explain structure download format - dotbracket is bad since it can't handle pseudoknots
    Explain DMS reactivity format

-->