<h1>Help</h1>

<h2>Genome browser</h2>
<p>
    FoldAtlas includes a genome browser called <i>d3nome</i>, which shows intron and exon boundaries for each transcript, along with features that are near to the transcript of interest. This facilitates genome exploration and identification of interesting transcripts.
</p>

[show a picture of the genome browser here]

[show bullet list describing each element]

<h2>Searching for transcripts</h2>
<h3>Go to a known transcript ID</h3>
<p>
    To search for a known transcript identifier (e.g. the TAIR identifier "AT1G01010.1"), hit <b>Search</b> in the title bar, followed by the <b>Transcript ID</b> tab. Next, the transcript transcript ID should be entered and the <b>Search</b> button pressed. This will search the FoldAtlas database and jump to the profile page of the first match.
</p>
<img class="help-example-image" src="{{ settings.static_base_url }}/assets/help/search-transcript-id.png" />

<h3>Find high coverage transcripts</h3>
<p>
    An issue with genome-scale RNA structure probing experiments is that many transcripts have only a small number of reads mapped to them, i.e. are low-coverage. We have produced an interface that lists transcripts in order of coverage. This can be reached by pressing <b>Search</b> and then choosing the <b>Coverage</b> tab. Transcripts with the best coverage are listed first, so that high-coverage transcripts can be found easily. The right column indicates transcripts where <i>in vivo</i> structures are available.
</p>
<img class="help-example-image" src="{{ settings.static_base_url }}/assets/help/search-coverage.png" />

<h2>Structure views</h2>
<p>
    FoldAtlas holds RNA secondary structure predictions generated using the RNAstructure package [REF]. Predictions are divided into two categories: <i>in vivo</i>, and <i>in silico</i>. Each structure prediction has an estimated free energy term associated, in kcal/mol.

<h3><i>In silico</i> predictions</h3>
<p>
    All transcript sequences have <i>in silico</i> structures, which have been generated from the RNA sequence alone, without constraints.
</p> 

<h3><i>In vivo</i> predictions</h3>
<p>
    The <i>in vivo</i> structures have been generated by constraining the prediction method with dimethyl sulfate (DMS) reactivity values, generated in a previous genome-scale screen [REF]. Constraining predictions using experimental structure probing data has been shown to improve the predictive accuracy [REF], hence <i>in vivo</i> predictions are inferred to be of better accuracy than their <i>in silico</i> counterparts.
</p><p>
    It is important to note that not all transcripts have <i>in vivo</i> predictions, since many transcripts have low read sequence coverage, corresponding to low RNA abundance. The proportion of transcripts with <i>in vivo</i> predictions will improve in future FoldAtlas iterations, as read depth from experiments increases.
<p>

<h3>PCA plots for structural similarity visualisation</h3>
<img class="help-example-image-small" src="{{ settings.static_base_url }}/assets/help/pca.png" />
<p>
    When the RNAstructure prediction method is run on a transcript sequence, up to 20 low free energy structures are generated. In FoldAtlas, the similarity between these structures is visualised using Principle Component Analysis (PCA). Each dot on the plot is a structure, and similar structures are located in close proximity to one another.
</p><p>
    Clicking a dot shows detailed information about the structure. A red dot colour indicates lower free energy, i.e. a more favourable structure. The minimum free energy (MFE) structure is visualised by default.
</p><p>
    The method for using PCA with RNA structures has been described in an earlier work [REF], and involves generating a vector of length <i>n</i>, where <i>n</i> is the number of nucleotides. The value at each position in the vector is set to either 0 if the corresponding base is unpaired, or 1 if paired. The vectors are then entered into the PCA method [REF], with each vector position as a dimension.
</p>

<h3>Circle plots for base pairing visualisation</h3>
<img class="help-example-image-small" src="{{ settings.static_base_url }}/assets/help/circle.png" />
<p>
    The circle plot can be used to intuitively visualise base pairing in a predicted RNA structure [REF]. Bases are arranged in a circle according to their sequence position, and lines drawn between bases indicate pairing. The free energy estimate is displayed in the top right corner.
</p>

<h3>Structure diagrams for fold visualisation</h3>
<img class="help-example-image-small" src="{{ settings.static_base_url }}/assets/help/structure-diagram.png" />
<p>
    FoldAtlas also visualises structures using a fold diagram. The diagram is generated using the <i>ViennaRNA</i>[REF] package and then displayed in the browser using the <i>forna</i>[REF] JS module. Clicking </b>Enable interaction</b> allows the structure diagram to be rearranged by clicking and dragging bases. A free energy estimate is displayed in the top right corner.
</p><p>
    The nucleotides are coloured according to their reactivity. Green nucleotides represent weak or no reactivity, orange denotes moderate reactivity, whilst red indicates strong reactivity. White nucleotides indicate that reactivity is irrelevant. G and U nucleotides are always coloured white because we are displaying DMS reactivity data, and DMS only reacts with unpaired C and A nucleotides.
</p>

<h3>Structure data download</h3>
<img class="help-example-image" src="{{ settings.static_base_url }}/assets/help/structure-download.png" />
<p>
    The <b>Download</b> button illustrated above generates a tab-delimited text file containing up to 20 <i>in silico</i> structures, along with up to 20 <i>in vivo</i> structures if coverage is sufficient. The following snippet illustrates 10 bases from an <i>in silico</i> structure.
</p>

<pre class="help-pre">
<u>1       2          3       4           5       6       7       8   9   10 </u>
2029    In silico  Col_0   AT1G01620.1 -394.1  4.55739 2.90891 A   35  0
2029    In silico  Col_0   AT1G01620.1 -394.1  4.55739 2.90891 A   36  0
2029    In silico  Col_0   AT1G01620.1 -394.1  4.55739 2.90891 A   37  0
2029    In silico  Col_0   AT1G01620.1 -394.1  4.55739 2.90891 C   38  134
2029    In silico  Col_0   AT1G01620.1 -394.1  4.55739 2.90891 C   39  133
2029    In silico  Col_0   AT1G01620.1 -394.1  4.55739 2.90891 A   40  0
2029    In silico  Col_0   AT1G01620.1 -394.1  4.55739 2.90891 T   41  132
2029    In silico  Col_0   AT1G01620.1 -394.1  4.55739 2.90891 T   42  131
2029    In silico  Col_0   AT1G01620.1 -394.1  4.55739 2.90891 C   43  130
2029    In silico  Col_0   AT1G01620.1 -394.1  4.55739 2.90891 T   44  129
2029    In silico  Col_0   AT1G01620.1 -394.1  4.55739 2.90891 T   45  0
</pre>

<p>
    The columns represent, from left to right:

    <ol>
        <li>A unique identifier for the structure, used internally by FoldAtlas</li>
        <li>Description of the prediction method; can be <i>in silico</i> or <i>in vitro</i>.</li>
        <li>The strain from which the RNA sequence originated. Currently all structures are for Col_0, which is the <i>A. thaliana</i> reference strain.</li>
        <li>The TAIR transcript indentifier.</li>
        <li>Free energy estimate, in kcal/mol, generated by <i>RNAstructure</i>.</li>
        <li>Principle Component 1, calculated during PCA. [INTERNAL REF]</li>
        <li>Principle Component 2, calculated during PCA. [INTERNAL REF]</li>
        <li>Nucleotide letter.</li>
        <li>Position of this nucleotide.</li>
        <li>The position that this nucleotide pairs to. 0 indicates no base pairing.</li>
    </ol>
</p>

<h2>DMS reactivity measurements</h3>
<p>
    FoldAtlas includes a module that plots DMS reactivity charts and allows them to be downloaded. At present, all FoldAtlas structure probing measurements are from DMS experiments. In the future, we will also add selective 2'-hydroxyl acylation analyzed by primer extension (SHAPE) measurements. 

    EXPLAIN NORMALISATION
</p>

<h3>DMS measurement charts</h3>
<p>
    The chart area includes an overview mode, which displays reactivities for the entire transcript on a single line. This can be seen at the top of the image below. Underneath the overview is a <b>More detail</b> button, which shows the transcript on multiple lines, with 80 nucleotides per line. In this detailed view, G and U nucleotide labels are coloured grey to indicate that they do not react with DMS.
</p>
<img class="help-example-image" src="{{ settings.static_base_url }}/assets/help/measurements-chart.png" />

<h3>Downloading DMS measurements</h3>
<img class="help-example-image" src="{{ settings.static_base_url }}/assets/help/dms-download.png" />
<p>
   The <b>Download</b> button can be used to obtain a tab-delimited text file containing the DMS reactivity measurements for a given transcript. The following snippet illustrates 10 bases from the AT1G01620.1 transcript:
</p>

<pre class="help-pre">
<u>1   2   3       </u>
75  U   NA
76  G   NA
77  U   NA
78  A   0.492188
79  G   NA
80  A   0.767697
81  A   0.940973
82  A   0.45815
83  A   0.0
84  G   NA
85  A   0.823105
</pre>

<p>
    The columns represent, from left to right:

    <ol>
        <li>The nucleotide position. Numbers start from 1.</li>
        <li>The base type letter.</li>
        <li>The reactivity. 0 indicates no reactivity whilst 1 or more indicates the strongest reactivity. NA indicates a G or U nucleotide, where DMS reactivities are not relevant.</li>
    </ol>
</p>

<h2>References</h2>

<!--

Genome browser 
    Click and drag to move 
    Use menu to select chromosome
    Click labels to see details

References

DMS reactivities
    Source of data
    Spliced transcripts
    Description of the normalisation method
    Detailed view

(Ribosome profiling - later)

Downloading data
    Explain Download All SQL format, including ER diagram
    Explain structure download format - dotbracket is bad since it can't handle pseudoknots
    Explain DMS reactivity format

-->