<h1>Help</h1>

<h2>Genome browser</h2>
<p>
	FoldAtlas includes a genome browser called d3nome, which shows intron and exon boundaries for each transcript, along with features that are near to the transcript of interest. This facilitates genome exploration and identification of interesting transcripts.
</p>

[show a picture of the genome browser here]

[show bullet list describing each element]

<h2>Searching for transcripts</h2>
<h3>Go to a known transcript ID</h3>
<p>
	To search for a known transcript identifier (e.g. the TAIR identifier "AT1G01010.1"), hit "Search", followed by the "Transcript ID" tab. Next, the transcript transcript ID should be entered and the "Search" button pressed. This will search the FoldAtlas database for the transcript ID, and jump to the profile page of the first match.
</p>
<img class="help-example-image" src="{{ settings.static_base_url }}/assets/help/search-transcript-id.png" />

<h3>Find high coverage transcripts</h3>
<p>
	An issue with genome-scale RNA structure probing experiments is that many transcripts have only a small number of reads mapped to them, i.e. are low-coverage. We have produced an interface that lists transcripts in order of coverage. This can be reached by pressing Search and then choosing the Coverage tab. Transcripts with the best coverage are listed first, so that high-coverage transcripts can be found easily. The right column indicates transcripts where <i>in vivo</i> structures are available.
</p>
<img class="help-example-image" src="{{ settings.static_base_url }}/assets/help/search-coverage.png" />

<h3>Structure view</h3>
<p>
	FoldAtlas holds RNA secondary structure predictions generated using the RNAstructure package [REF]. Predictions are divided into two categories: <i>in vivo</i>, and <i>in silico</i>. Each structure prediction has an estimated free energy term associated, in kcal/mol.

<h4><i>In silico</i></h4>
<p>
	All transcript sequences have <i>in silico</i> structures, which have been generated from the RNA sequence alone, without constraints.
</p> 

<h4><i>In vivo</i></h4>
<p>
	The <i>in vivo</i> structures have been generated by constraining the prediction method with DMS reactivity values, generated in a previous genome-scale screen [REF]. Constraining predictions using experimental structure probing data has been shown to improve the predictive accuracy [REF]. It is important to note that not all transcripts have <i>in vivo</i> predictions, due to low read sequence coverage. The proportion of transcripts with <i>in vivo</i> predictions will improve in future versions, as read depth increases.
<p>

<h4>PCA plots</h4>

<img class="help-example-image-small" src="{{ settings.static_base_url }}/assets/help/pca.png" />

<p>
	When the RNAstructure prediction method is run on a transcript sequence, up to 20 low free energy structures are generated. In FoldAtlas, the similarity between these structures is visualised using Principle Component Analysis (PCA). Each dot on the plot is a structure, and similar structures are located in close proximity to one another. Clicking a dot shows detailed information about the structure. The free energy of each structure is also visualised according to the dot colour, with red indicating lower free energy.
</p>

<p>
	The method for using PCA with RNA structures has been described in an earlier work [REF], and involves generating a vector of length <i>n</i>, where <i>n</i> is the number of nucleotides. The value at each position in the vector is set to either 0 if the corresponding base is unpaired, or 1 if paired. The vectors are then entered into the PCA method [REF], with each vector position as a dimension.
</p>

</p>

<h3>References</h3>



<!--

	Transcript ID - 
	Search by coverage

Genome browser 
	Click and drag to move 
	Use menu to select chromosome
	Click labels to see details

Structure view
	PCA
	Circle plot
	Structure diagram

DMS reactivities
	Source of data
	Spliced transcripts
	Description of the normalisation method
	Detailed view

(Ribosome profiling - later)

Downloading data
	Explain Download All SQL format, including ER diagram
	Explain structure download format - dotbracket is bad since it can't handle pseudoknots
	Explain DMS reactivity format

-->