<h1>Help</h1>

<h2>Genome browser</h2>
<p>
	FoldAtlas includes a genome browser called d3nome, which shows intron and exon boundaries for each transcript, along with features that are near to the transcript of interest.

	
</p>

[show a picture of the genome browser here]

[show bullet list describing each element]

<h2>Searching for transcripts</h2>
<h3>Go to a known transcript ID</h3>
<p>
	To search for a known transcript identifier (e.g. the TAIR identifier "AT1G01010.1"), the "Search" button should be pressed, followed by the "Transcript ID" tab. Next, the transcript transcript ID should be entered and the "Search" button pressed. This will search the FoldAtlas database for the transcript ID, and jump to the profile page of the first match.
</p>

<h3>Find high coverage transcripts</h3>
<p>
	An issue with genome-scale RNA structure probing experiments is that many transcripts have only a small number of reads mapped to them, i.e. are low-coverage. We have produced an interface that lists transcripts in order of coverage. This can be reached by pressing Search and then choosing the Coverage tab. Transcripts with the best coverage are listed first, so that high-coverage transcripts can be found easily. 
</p>

<!--

	Transcript ID - 
	Search by coverage

Genome browser 
	Click and drag to move 
	Use menu to select chromosome
	Click labels to see details

Structure view
	PCA
	Circle plot
	Structure diagram

DMS reactivities
	Source of data
	Spliced transcripts
	Description of the normalisation method
	Detailed view

(Ribosome profiling - later)

Downloading data
	Explain Download All SQL format, including ER diagram
	Explain structure download format - dotbracket is bad since it can't handle pseudoknots
	Explain DMS reactivity format

-->